### PR TITLE
Refactor combining session and local storage proxy

### PR DESCRIPTION
**Pull request by mrclay**

### CODE REVIEW

File: src/crtoolLocalStorage.js [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+        // If review_id found but a token is not set, update the URL
+        if (!token && review_id) {
+            const new_url = ls.updateUrlParameter(window.location.href, 'token', review_id);
+            window.history.pushState({ path: new_url }, '', new_url);
+        }
+
+        let db_review;
+        try {
+            db_review = await ls.fetchReviewFromServer(review_id);
```
Comment (mrclay) [inline src/crtoolLocalStorage.js:56]: Like loadReviewFromLocalStorage, this should return undefined if no valid review is returned, but throw on a request failure.

File: src/crtoolLocalStorage.js [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            // Do not allow end users to set their own ID via the URL or localStorage.
+            localStorage.removeItem('crtool.' + review_id);
+            localStorage.removeItem('curriculumReviewId');
+            redirectHome();
+            return;
+        }
+
+        const ls_review = ls.loadReviewFromLocalStorage(review_id);
+
+        if (ls_review && ls_review.ls_modified_time > db_review.ls_modified_time) {
```
Comment (mrclay) [inline src/crtoolLocalStorage.js:78]: I didn't see a good reason to test last_updated anymore, but maybe I didn't fully capture the cases we're handling.

File: src/crtoolLocalStorage.js [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        ls.review.ls_modified_time = (new Date()).toUTCString();
+        localStorage.setItem('crtool.' + ls.review.id, JSON.stringify(ls.review));
+    },
+
+    isValidReview(review) {
+        if (typeof review !== 'object' || review === null) {
+            return false;
+        }
+
+        for (const prop of ['id', 'last_updated', 'ls_modified_time']) {
```
Comment (nbucknor) [inline src/crtoolLocalStorage.js]: @mrclay reviews are initially created in python. ls_modified_time is created in JavaScript. So this property will not exist the first time a review is created and therefore will always fail the isValidReview() test.

File: src/crtoolLocalStorage.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+                }
+
+                if (xhttp.status !== 200) {
+                    console.error('Review fetch failed.');
+                    reject();
+                }
+
+                try {
+                    const review = JSON.parse(xhttp.responseText);
+                    if (!ls.isValidReview(review)) {
```
Comment (nbucknor) [inline src/crtoolLocalStorage.js]: This will always fail the first time a review is created.

File: src/crtoolLocalStorage.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+        let db_review;
+        try {
+            db_review = await ls.fetchReviewFromServer(review_id);
+        } catch (e) {
+            console.error(e);
+            // Do not allow end users to set their own ID via the URL or localStorage.
+            localStorage.removeItem('crtool.' + review_id);
+            localStorage.removeItem('curriculumReviewId');
```
Comment (nbucknor) [inline src/crtoolLocalStorage.js]: Will this mean that the localStorage version of the review will get deleted if any type of error occurs while trying to load it from the server? Server timeouts, lost internet connection, etc?

**Comment by mrclay** [comment]
I addressed the bugs and added commits to fix the existing unit test suite, and make sure saves are more safe (last commit).

**Pull request merged**